clippy_lints/src/to_string_in_display.rs

```diff
@@ -0,0 +1,79 @@
+use crate::utils::{match_def_path, match_trait_method, paths, span_lint};
+use if_chain::if_chain;
+use rustc_hir::*;
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for uses of `to_string()` in `Display` traits.
+    ///
+    /// **Why is this bad?** Usually `to_string` is implemented indirectly
+    /// via `Display`. Hence using it while implementing `Display` would
+    /// lead to infinite recursion.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    ///
+    /// ```rust
+    /// use std::fmt;
+    ///
+    /// struct Structure(i32);
+    /// impl fmt::Display for Structure {
+    ///     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+    ///         write!(f, "{}", self.to_string())
+    ///     }
+    /// }
+    ///
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// use std::fmt;
+    ///
+    /// struct Structure(i32);
+    /// impl fmt::Display for Structure {
+    ///     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+    ///         write!(f, "{}", self.0)
+    ///     }
+    /// }
+    /// ```
+    pub TO_STRING_IN_DISPLAY,
+    correctness,
+    "to_string method used while implementing Display trait"
+}
+
+declare_lint_pass!(ToStringInDisplay => [TO_STRING_IN_DISPLAY]);
+
+impl LateLintPass<'_> for ToStringInDisplay {
+    fn check_expr(&mut self, cx: &LateContext<'_>, expr: &Expr<'_>) {
+        let parent_id = cx.tcx.hir().get_parent_item(expr.hir_id);
+        let parent_node = cx.tcx.hir().find(parent_id);
+
+        if_chain! {
+            if let ExprKind::MethodCall(ref path, _, _, _) = expr.kind;
+            if path.ident.name == sym!(to_string)
+                    && match_trait_method(cx, expr, &paths::TO_STRING);
+            if path.ident.as_str() == "to_string";
+            if let Some(Node::ImplItem(ImplItem {ident, kind, ..})) = parent_node;
+            if ident.as_str() == "fmt";
+            if let ImplItemKind::Fn(FnSig {decl, ..}, _) = kind;
+            if let FnDecl{inputs, ..} = decl;
+            if let Ty {kind, ..} = &inputs[0];
+            if let TyKind::Rptr(_, MutTy {ty, ..}) = kind;
+            if let Ty {kind, ..} = ty;
+            if let TyKind::Path(QPath::Resolved(_, path)) = kind;
+            if let Path{res, ..} = path;
+            if let def::Res::SelfTy(Some(did), _) = res;
+            if match_def_path(cx, *did, &paths::DISPLAY_TRAIT);
```

Comment:
You can remove this whole code if you add:

```rust
fn is_display_impl(cx: ..., item: &Item<'_>) {
    if_chain! {
        if let ItemKind::Impl { of_trait: Some(trait_ref), .. } = item.kind;
        if let Some(did) = trait_ref.trait_def_id();
        if match_def_path(cx, did, &paths::DISPLAY_TRAIT);
        then {
            true
        } else {
            false
        }
    }
}

fn check_item(&mut self, cx: ..., item: &Item<'_>) {
    if is_display_impl(item) {
        self.in_display_impl = true;
    }
}

fn check_item_post(&mut self, cx: ..., item: &Item<'_>) {
    if is_display_impl(item) {
        self.in_display_impl = false;
    }
}
```

and then

```suggestion
            if self.in_display_impl;
```

Reply:
I'm trying below code but returns `this if-then-else expression returns a bool literal` error. 

```
fn is_display_impl(cx: &LateContext<'_>, item: &Item<'_>) -> bool {
    if_chain! {
        if let ItemKind::Impl { of_trait: Some(trait_ref), .. } = &item.kind;
        if let Some(did) = trait_ref.trait_def_id();
        if match_def_path(cx, did, &paths::DISPLAY_TRAIT);
        then {
            true
        } else {
            false
        }
    }
}
```